### PR TITLE
Reputation 2.4

### PR DESCRIPTION
**Pull request by HitoriWoB**
Ready for dev testing (reputation decay)

### CODE REVIEW

File: classes/Reputation.php
Comment (levimeahan) [inline classes/Reputation.php:96]: Instead of string interpolation, what if we made this a nested array? Seems a little cleaner and easier to do static type checking
```suggestion
    const DAILY_TASK_GAINS = [
        DailyTask::DIFFICULTY_EASY => [
            DailyTask::ACTIVITY_ARENA => 1,
        ],
        DailyTask::DIFFICULTY_MEDIUM => [
            DailyTask::ACTIVITY_ARENA => 3,
        ],
        DailyTask::DIFFICULTY_HARD => [
            DailyTask::ACTIVITY_ARENA => 5,
        ],
```

File: classes/Reputation.php
Comment (levimeahan) [inline classes/Reputation.php:3]:
```suggestion
class UserReputation {
```

**Review by levimeahan — APPROVED** [review]
LGTM! Could do a couple minor fixes before or after merging

**Comment by HitoriWoB** [comment]
image
Dev bullet points (will expand into an actual news post once you've approved changes)

**Comment by HitoriWoB** [comment]
Have PvP rep and one more bug patch. Please hold off on requested review

File: classes/User.php
Comment (levimeahan) [inline classes/User.php:977]: Relying on this method being called for `trainingManager` to be present introduces a lot of bug possibilities where you call `$player->trainingManager` but it wasn't loaded. We could fix this in a couple ways:

1) Always instantiate this when the user class is loaded (call this method in `loadData()`)
2) Make the property private and make this method `getTrainingManager()`. Inside this we can return `$this->trainingManager` if already initialized, if not we initialize first.

File: classes/User.php
Comment (levimeahan) [inline classes/User.php:1698]:
```suggestion
		`recent_players_killed_ids` = '$this->recent_players_killed_ids',
```

File: classes/UserReputation.php
Comment (levimeahan) [inline classes/UserReputation.php:132]:
```suggestion
    const PVP_CHAIN_KILL_LIMIT = 4; // Pvp kills of same player above this amount reward 0 rep
```

File: classes/UserReputation.php
Comment (levimeahan) [inline classes/UserReputation.php:131]:
```suggestion
    const RECENT_PLAYER_KILL_THRESHOLD = 3600; // Only kills within last hour will mitigate pvp rep gains
```

**Review by levimeahan — APPROVED** [review]

File: classes/UserReputation.php
Comment (levimeahan) [inline classes/UserReputation.php:291]: This is backwards, below the comments state that `$level_difference = -5` means that the opponent is 5 levels below player.

However if player level is 75 and opponent level is 70, the value of `$level_difference` is 75 - 70, or +5

Comment (levimeahan) [inline classes/UserReputation.php:291]:
```suggestion
        $player_levels_above_opponent = $opponent_level - $player_level;
```


This is backwards, below the comments state that $level_difference = -5 means that the opponent is 5 levels below player.

However if player level is 75 and opponent level is 70, the value of $level_difference is 75 - 70, or +5

**Pull request merged**
